Source/JavaScriptCore/API/glib/JSCValue.cpp

```diff
@@ -2145,3 +2145,85 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
 
     return json;
 }
+
+/**
+ * JSCExecutor:
+ * @resolve: #JSCValue function to call to resolve the promise
+ * @reject: #JSCValue function to call to reject the promise
+ * @user_data: user data
+ *
+ * A function passed to @jsc_value_new_promise called during initialization
+ *
+ * It is called like a JavaScript function, so exceptions raised will not be propagated
+ * to the context, but handled by the promise causing a rejection.
+ * @resolve and @reject can be reffed for later use to handle async task completion.
+ *
+ * Since: 2.48
+ */
+
+/**
+ * jsc_value_new_promise:
+ * @context: a #JSCContext
+ * @executor: (scope call) (closure user_data): an initialization callback
+ * @user_data: (nullable): user data passed in @executor
+ *
+ * Creates a new Promise. @executor will be invoked during promise initialization
+ * and it receives the @resolve and @reject objects than can be called to resolve
+ * or reject the promise. It is called like a JavaScript function, so exceptions raised
+ * during the executor invocation will not be propagated to the context, but
+ * handled by the promise causing a rejection.
+ *
+ * Here is an example usage:
+ * |[<!-- language="C" -->
+ * // An asynchronous task result
+ * static gboolean async_response_callback(gpointer userData)
+ * {
+ *     Foo* foo = (Foo*)(userData);
+ *     g_object_unref(jsc_value_function_call(foo->resolve, G_TYPE_NONE));
+ *     g_object_unref(foo->resolve);
+ *     return FALSE;
+ * }
+ * // The executor required to hold a ref on resolve or reject function
+ * static void fooExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
+ * {
+ *     Foo* foo = (Foo*)(userData);
+ *     foo->resolve = g_object_ref(resolve);
+ *     g_timeout_add(200, async_response_callback, userData);
+ * }
+ * // A registered JavaScript function or method
+ * static JSCValue* getAsyncJobPromise(Foo* foo)
+ * {
+ *     return jsc_value_new_promise(jsc_context_get_current(), fooExecutor, foo);
+ * }
+ * ]|
```

Comment:
I'm not sure this example is very clear to be honest, since it's copied from the unit tests without context about what Foo is. I don't want to block this more, because the implementation and unit tests are ok. So, let's remove the code example for now, we can improve the docs later in a follow up.